iOS web push processing before unlock can cause InvalidStateError exceptions

bnham opened the pull request:
#### 6eefd2796ca6a2b650c4d0caed20137bbdb91bbd
<pre>
iOS web push processing before unlock can cause InvalidStateError exceptions
<a href="https://bugs.webkit.org/show_bug.cgi?id=272865">https://bugs.webkit.org/show_bug.cgi?id=272865</a>
<a href="https://rdar.apple.com/126061835">rdar://126061835</a>

Reviewed by Brady Eidson.

If a web push is processed before the user unlocks the device for the very first time, then that
processing fails because the PushDatabase file has file protection attributes that prevent it from
being opened before first unlock. This then causes the PushService to get in to a state where it
responds to all IPCs with an InvalidStateError since the push database couldn&apos;t be initialized.

Changing the protection class to always allow access is probably not a great idea because no one has
ever validated whether all the downstream work involved in handling a push event (e.g. spawning a
service worker in a new process) works before first unlock. It likely doesn&apos;t, because many files on
the volume will have a protection class attribute that makes them inaccessible before first unlock.

To fix this, hold off on database initialization until the user unlocks the device for the first
time. It&apos;s important that we only hold off only on DB initialization and not on initializing the
push service connection (i.e. creating ApplePushServiceConnection) because apsd needs us to check in
promptly or else it will put our daemon in a penalty box for not responding to its launch on demand
request promptly.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/spi/ios/MobileKeyBagSPI.h: Added.
* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.webpushd.sb.in:
* Source/WebKit/webpushd/PushService.mm:
(WebPushD::performAfterFirstUnlock):
(WebPushD::hasUnlockedAtLeastOnce):
(WebPushD::PushService::create):
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/MobileKeyBag.framework/MobileKeyBag.tbd: Added.

Canonical link: <a href="https://commits.webkit.org/277859@main">https://commits.webkit.org/277859@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c9e553fe10e28564b426b2c1e3fdf1d2a33876d6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48813 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28025 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/51779 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51499 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44879 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51118 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/33969 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25554 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/39902 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/49395 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25672 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/42090 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21003 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23135 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/43273 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6868 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/42112 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/45066 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/43760 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/53410 "Built successfully") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/48304 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/23864 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/20133 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47209 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/25127 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42298 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/46153 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10743 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25934 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/55799 "Built successfully") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/24847 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11487 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

beidson commented on Source/WebKit/webpushd/PushService.mm:
```diff
@@ -55,32 +61,69 @@ static void updateTopicLists(PushServiceConnection& connection, PushDatabase& da
+    static std::once_flag once;
+    static NeverDestroyed<Vector<Function<void()>>> functions;
+
+    std::call_once(once, [function = WTFMove(function)]() mutable {
```
This does a WTFMove() of the function...

(See below)

bnham replied:
Good catch, I left this in from an earlier version of the patch and it's not necessary now (and will cause a crash).

beidson commented on Source/WebKit/webpushd/PushService.mm:
```diff
@@ -55,32 +61,69 @@ static void updateTopicLists(PushServiceConnection& connection, PushDatabase& da
+        }).get());
+    });
+
+    functions->append(WTFMove(function));
```
And this does a WTFMove() of the function.

(See above)

For the first function that comes in, that's not right, is it?
(I might be missing something)

beidson reviewed (commented):
Seems fine *except* for the possible double WTFMove() of the function I commented on?

bnham commented on WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/MobileKeyBag.framework/MobileKeyBag.tbd:
```diff
@@ -0,0 +1,8 @@
+--- !tapi-tbd
+tbd-version: 4
+targets: [armv7-ios, armv7s-ios, arm64-ios, arm64e-ios]
+install-name: /System/Library/PrivateFrameworks/CorePrediction.framework/CorePrediction
```
oops

beidson commented on Source/WebKit/webpushd/PushService.mm:
```diff
@@ -55,32 +61,69 @@ static void updateTopicLists(PushServiceConnection& connection, PushDatabase& da
+        }).get());
+    });
+
+    functions->append(WTFMove(function));
```
On the first call to `performAfterFirstUnlock`, you changed the lambda capture list on line 77 to not WTFMove `function`, but `makeBlockPtr` still grabs it.

This is (bizarrely and problematically) still allowed because of how Obj-C blocks work.

The capture is simply not needed, because the function that's setup inside walks the `functions` Vector and doesn't care about the initial `function` argument.

bnham replied:
Yep, I actually submitted a new version of the PR to fix this in the same minute as this comment.

bnham committed:
iOS web push processing before unlock can cause InvalidStateError exceptions
https://bugs.webkit.org/show_bug.cgi?id=272865
rdar://126061835

Reviewed by Brady Eidson.

If a web push is processed before the user unlocks the device for the very first time, then that
processing fails because the PushDatabase file has file protection attributes that prevent it from
being opened before first unlock. This then causes the PushService to get in to a state where it
responds to all IPCs with an InvalidStateError since the push database couldn't be initialized.

Changing the protection class to always allow access is probably not a great idea because no one has
ever validated whether all the downstream work involved in handling a push event (e.g. spawning a
service worker in a new process) works before first unlock. It likely doesn't, because many files on
the volume will have a protection class attribute that makes them inaccessible before first unlock.

To fix this, hold off on database initialization until the user unlocks the device for the first
time. It's important that we only hold off only on DB initialization and not on initializing the
push service connection (i.e. creating ApplePushServiceConnection) because apsd needs us to check in
promptly or else it will put our daemon in a penalty box for not responding to its launch on demand
request promptly.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/spi/ios/MobileKeyBagSPI.h: Added.
* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.webpushd.sb.in:
* Source/WebKit/webpushd/PushService.mm:
(WebPushD::performAfterFirstUnlock):
(WebPushD::hasUnlockedAtLeastOnce):
(WebPushD::PushService::create):
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/MobileKeyBag.framework/MobileKeyBag.tbd: Added.

Canonical link: https://commits.webkit.org/277859@main
